### ISSUE TITLE
_xsrf cookie for singleuser server not scoped to the base_url

### ISSUE THREAD

**Issue by albertmichaelj** [jupyterhub/jupyterhub#3624]
EDIT: The issue with the `jupyterhub-session-id` cookie has been fixed with pull request #3625, which has been merged in master. The issue with the `_xsrf` cookie remains, though there are configuration changes that can be made to fix it.

### Bug description

JupyterHub has the ability to be accessed at a subpath of a domain (such as `www.whatever.com/jupyterhub`) using the configurable `c.JupyterHub.base_url="/jupyterhub"`. When this is done, most of the cookies that JupyterHub uses are effectively scoped to the path "/jupyterhub/" (or a further sub-path). However, the cookies `jupyterhub-session-id` and `_xsrf` are scoped to the path `/`. This is maybe a problem, but I actually haven't figured out if it actually causes bugs. The potential issue is that if you run two instances of JupyterHub at two different subpaths of the same domain (i.e. `www.whatever.com/jupyterhub1` and `www.whatever.com/jupyterhub2`) the `jupyterhub-session-id` and `_xsrf` cookies are effectively shared between the two instances if you are logged into both instances at the same time. Whichever server wrote the cookie first gets to dictate the cookie value. Moreover, if you log out of one of the servers, it will delete the `jupyterhub-session-id` and `_xsrf` cookies even if you are still logged into the other server.

I actually do run multiple instances of Jupyterhub on sub-paths of a single domain, and that is how I noticed the cookie issue. However, I have been unable to determine if anything breaks because of this. As far as I can tell, even though `jupyterhub-session-id` and `_xsrf` occasionally gets mistakenly deleted due to above issue, the other server keeps working fine. However, this is likely unintended behavior, so I wanted to bring it up.

#### Expected behaviour

I expect all cookies to be scoped to the subpath indicated by `base_url`.

#### Actual behaviour

This doesn't happen for the cookies `jupyterhub-session-id` and `_xsrf`

### How to reproduce

Run a jupyterhub server with the following command:

```
jupyterhub --JupyterHub.base_url="/jupyterhub"
```

Log in at `http://localhost:8000/jupyterhub`. Then look at the cookies and the paths for the cookies (browser dependent as to how to do this).

### Proposed Solution
I have already forked JupyterHub and implemented a two line fix for `jupyterhub-session-id`. I will create a separate pull request and link it back to this issue after submitting. Fixing `jupyterhub-session-id` seems very straightforward and, I think, uncontroversial.

However, the `_xsrf` cookie is trickier. This is because the cookie is actually set by the notebook server. The path issue can be fixed by changing both the config file for JupyterHub and for the notebook server as follows:

```python
c.JupyterHub.tornado_settings = dict(xsrf_cookie_kwargs=dict(path="/jupyterhub/"))
```

and

```python
c.ServerApp.tornado_settings = dict(xsrf_cookie_kwargs=dict(path="/jupyterhub/"))
```

Note: Replace `ServerApp` with `NotebookApp` if not using `jupyter_server`.

However, it's not enough to change this configuration in one instance because the notebook server sets the cookie and therefore sets the path, but JupyterHub deletes the cookie, and must know the path in order to delete the cookie. Without some communication between the notebook server and JuptyerHub, I'm not sure how to resolve this. Nor am I sure the best way to perform that communication.

I'm happy to try to work on a solution for that cookie if someone who understand the project a bit better would suggest a potential way to communicate the necessary configuration (`xsrf_cookie_kwargs`) between the two servers.

**Comment by manics** [comment]
It's fine to keep this issue, if you want you can edit the title to make it clear it's just one issue left.

The removal of the `_xsrf` cookie was added in https://github.com/jupyterhub/jupyterhub/pull/3341

Should notebook and hub be using the same xsrf cookie though, or should they use different paths e.g. `/baseurl` for JupyterHub, `/baseurl/user/<username>` for the notebook server? This is beyond my knowledge, I think @minrk is the best person to answer.

**Comment by albertmichaelj** [comment]
Sounds good about keeping the issue. I edited the title, and I added a note at the top to make it immediately obvious that part of the text in the issue is not outdated.

I don't actually think that the hub uses the `_xsrf` cookie at all. As far as I can tell the `xsrf_cookies` setting for tornado needs to be set to `True` for the hub to use `_xsrf` cookies, and it doesn't seem like it is set for JupyterHub (at least by default). Moreover, no matter what I do on my hub setup, an `_xsrf` cookie is not created until I start a notebook server. In fact, if I set `xsrf_cookies=True` in the tornado settings, I actually get the following error in the logs when trying to log in:
```
[W 2021-09-28 11:23:53.088 JupyterHub web:1787] 403 POST /hub/login?next=%2Fhub%2Fhome (::1): '_xsrf' argument missing from POST
[E 2021-09-28 11:23:53.088 JupyterHub web:1219] Uncaught exception in write_error
    Traceback (most recent call last):
      File "/Users/Michael/.brew/Caskroom/miniconda/base/envs/jupyterhub_dev/lib/python3.9/site-packages/tornado/web.py", line 1217, in send_error
        self.write_error(status_code, **kwargs)
      File "/Users/Michael/Documents/Temp/jupyterhub/jupyterhub/handlers/base.py", line 1310, in write_error
        html = self.render_template('%s.html' % status_code, sync=True, **ns)
      File "/Users/Michael/Documents/Temp/jupyterhub/jupyterhub/handlers/base.py", line 1223, in render_template
        template_ns.update(self.template_namespace)
      File "/Users/Michael/Documents/Temp/jupyterhub/jupyterhub/handlers/base.py", line 1233, in template_namespace
        user = self.current_user
      File "/Users/Michael/Documents/Temp/jupyterhub/jupyterhub/handlers/base.py", line 461, in current_user
        raise RuntimeError("Must call async get_current_user first!")
    RuntimeError: Must call async get_current_user first!
```

So, it looks like JuptyerHub is really just cleaning up after the notebook server, and the `_xsrf` token could really just be scoped to the user server subpath. It's not clear to me if the notebook server should instead just do the cleaning up, so maybe this is really something of a bug with `jupyter_server` (or jupyterlab).

Looking more carefully at the jupyterhub singleuser code, it seems like a number of things are passed in, including the base url. So, the issue may be that `jupyter_server` need to set the `xsrf_cookie_kwargs` correctly when starting the server, and then to do the clearing of the cookies themselves at logout. Since JupyterHub doesn't seem to use the `_xsrf` cookie, it's not clear why it's clearing it at all. In the future, if Jupyterhub does want to use the `_xsrf` cookie for some reason, it would be better to scope the cookie to the `base_url/hub` subpath so it is distinct from the notebook app server anyway.

However, I'm still pretty new to the code here, so I don't want to assume I've understood everything correctly. If someone who understands things better can confirm that JuptyerHub indeed does not use the `_xsrf` cookie at all, then I would suggest reverting #3341 in favor of opening a pull request to have `jupyter_server` correctly scope the creation of the `_xsrf` cookie and then delete it on logout.
